runtime/lua/vim/lsp/util.lua

```diff
@@ -1341,19 +1341,21 @@ do --[[ References ]]
   ---
   ---@param bufnr buffer id
   ---@param references List of `DocumentHighlight` objects to highlight
+  ---@param client_id number client id
   ---@see https://microsoft.github.io/language-server-protocol/specifications/specification-3-17/#documentHighlight
   function M.buf_highlight_references(bufnr, references, client_id)
     validate { bufnr = {bufnr, 'n', true} }
     local client = vim.lsp.get_client_by_id(client_id)
-    if not client then
-      return
+    local offset_encoding = "utf-16"
+    if client then
```

Comment:
I don't think we should set the global client level offset encoding via calling a util function as a side effect.

Reply:
What about something like:

```lua
function vim.lsp.util.get_client_offset_encoding(client_id)
  local client = vim.lsp.get_client_by_id(client_id)
  client and client.offset_encoding or "utf-16"
end
```

This isn't my codebase so I'm open to different ways to give the encoding a default value, but the existing code is a pretty rough way to introduce a breaking change (even if it's just nightly).

Reply:
(But this is a newly-introduced API on the development branch that very deliberately is not marked as stable -- breakage _must_ be expected.

That's not saying the API can't be improved, of course.)

Reply:
offset_encoding = client.offset_encoding or  "utf-16"

Reply:
> , but the existing code is a pretty rough way to introduce a breaking change (even if it's just nightly).

It's the nightly, breakages (and accidents) happen. You're free and encouraged to hit the watch button on the repo and review PRs into the nightly as a plugin author.

Reply:
Isn't a bit weird to have a "get client <foo>" function that has a fallback for _not_ passing a client?

Reply:
I was under the impression `vim.lsp.util.buf_highlight_references` is not a newly-introduced API, it exists even in `v0.5.0` with documentation, is there a list of stable APIs? The issue I have though is that this doesn't need to be a breaking change, when this gets released as stable it's gonna be an issue for every caller that exists (which is probably pretty small tbf) and I don't see how the average dev would fix their config/plugin without reading the source code.

Reply:
> offset_encoding = client.offset_encoding or "utf-16"

`client` is nilable.

Reply:
The docstring is indeed missing an update, and https://github.com/neovim/neovim/issues/14090 an update.

Reply:
> I was under the impression vim.lsp.util.buf_highlight_references is not a newly-introduced API, it exists even in v0.5.0 with documentation, is there a list of stable APIs? The issue I have though is that this doesn't need to be a breaking change, when this gets released as stable it's gonna be an issue for every caller that exists (which is probably pretty small tbf) and I don't see how the average dev would fix their config/plugin without reading the source code.

Ofc it's not, but this is the nightly we are talking about. You found and noticed an undocumented breaking change that can be worked around.

As for client being nillable, this is an lsp utility function, it's not meant to be used outside the lsp subsystem (that applies to all lsp utilities). Generally everything in the lsp subsystem assumes you are working within the lsp subsystem. If you're using this function for something *not* involving an active client (I have not looked at your plugin), then we can discuss moving it out of the lsp system and making it a general utility function.

Reply:
But stepping back here, making this client-dependent is _correct_; the change fixed actual bad behavior. And this is a utility function for the LSP client; if you (ab)use it for non-LSP purposes, that's neat but on you -- it doesn't make sense to "unfix" the function just for you!

Rather, what you need is a generic `vim.highlight` API that you (and LSP) can use. There's already such a module; maybe you can find a way to refactor things to make use of that?

Reply:
I'm actually tempted to move client out of this function and change the signature to directly take an offset encoding, then adjust the calling sites.

Reply:
> Ofc it's not, but this is the nightly we are talking about. You found and noticed an undocumented breaking change that can be worked around.

> The docstring is indeed missing an update, and #14090 an update.

I won't push this more, I get it it's nightly, addressing what clason mentioned is good enough.

> As for client being nillable, this is an lsp utility function, it's not meant to be used outside the lsp subsystem (that applies to all lsp utilities). Generally everything in the lsp subsystem assumes you are working within the lsp subsystem. If you're using this function for something not involving an active client (I have not looked at your plugin), then we can discuss moving it out of the lsp system and making it a general utility function.

I meant that `client` is nilable because existing callers of the function don't pass that argument in. FWIW the functionality of my plugin in question is LSP related.

Reply:
In the current implementation client_id should always be passed in, I'm in favor (if you're down, otherwise I'll do it), of adjust the calling functions to pass through the correct offset encoding because I do think it's a bit weird to pass the client_id into this function solely to fetch the offset encoding.

cc @mfussenegger and @elianiva

Reply:
> In the current implementation client_id should always be passed in, I'm in favor (if you're down, otherwise I'll do it), of adjust the calling functions to pass through the correct offset encoding because I do think it's a bit weird to pass the client_id into this function solely to fetch the offset encoding.

Sounds good to me

Reply:
> In the current implementation client_id should always be passed in, I'm in favor (if you're down, otherwise I'll do it), of adjust the calling functions to pass through the correct offset encoding because I do think it's a bit weird to pass the client_id into this function solely to fetch the offset encoding.

I agree it's better to do it this way.